### PR TITLE
Log epoch metrics before firing the `on_evaluation_end` hook

### LINKED ISSUE #7166: Log epoch metrics before the `on_evaluation_end` hook
## 🚀 Feature
I would like the epoch metrics to get logged before the `on_evaluation_end` hook.

### Motivation

When doing HPO with the [Ray Tune](https://docs.ray.io/en/master/tune/index.html) library, you usually report the metrics back to Tune `on_validation_end`. If you use a trial scheduler such as the ASHA algorithm, it will terminate the trial if the metric indicates low performance for this specific trial. If i want to visualize the HPO in WandB, for example, i would like to know for which values of the metric the trials got terminated. At the moment this is not possible, since the epoch metrics are not logged before the trial gets terminated.

### Pitch

Right now the order of hooks and logging is this one (taken from the `Trainer.run_evaluation()` method, version 1.3.0rc1):
```python
        # hook
        self.evaluation_loop.on_evaluation_epoch_end()

        # update epoch-level lr_schedulers
        if on_epoch:
            self.optimizer_connector.update_learning_rates(interval='epoch')

        # hook
        self.evaluation_loop.on_evaluation_end()

        # log epoch metrics
        eval_loop_results = self.evaluation_loop.log_epoch_metrics_on_evaluation_end()
```
Simply switching the order of the `log_epoch_metrics_on_evaluation_end` and the `on_evaluation_end` hook, would allow to inspect the epoch metrics in, for example, WandB, even if the trial gets terminated by Ray Tune. If one wants to create/modify the epoch metrics before they get logged, you still can use the `on_evaluation_epoch_end` hook.

### Alternatives

Introduce a new hook that is fired after the epoch metrics are logged. This hook could then be used by the tune reporter callback.

### Additional context

Not sure if there is an obvious reason for the current order, if so, please excuse my misunderstanding. 
If you think the change is desirable i would be more than happy to provide a pull request.

Keep up the great work, lightning is a really amazing library!

### PR DESCRIPTION
**Pull request by dcfidalgo**
## What does this PR do?

This PR switches the order in which the epoch metrics are logged and the `on_evaluation_end` hook is fired. Now it logs the metrics first. Fixes #7166

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by ananthsub** [comment]
Thanks for working on this @dcfidalgo ! Could you also add a test that confirms the log order?

**Comment by dcfidalgo** [comment]
Thanks @ananthsub for the feedback. Included a test and updated the changelog, let me know if this is ok like this.

**Comment by carmocca** [comment]
@dcfidalgo Mind resolving conflicts? This should be good to land now that 1.3 is released :)

**Comment by dcfidalgo** [comment]
@carmocca Done :)

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
as long as we don't allow logging on_evaluation_end() we should be good, right?

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1002,12 +1002,12 @@ def run_evaluation(self, on_epoch: bool = False) -> _EVALUATE_OUTPUT:
...
             )
 
-        # hook
-        self.evaluation_loop.on_evaluation_end()
-
         # log epoch metrics
         eval_loop_results = self.logger_connector.get_evaluate_epoch_results()
 
+        # hook
+        self.evaluation_loop.on_evaluation_end()
```
Comment (tchaton) [inline pytorch_lightning/trainer/trainer.py:1009]: Some users might be logging in on_evaluation_end and this a BC. Also did you disable the logging in `on_evaluation_end` hook ?

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:1009]: `on_evaluation_end` calls either `on_test_end` or `on_validation_end` and they do not support logging (in current master)

See: https://github.com/PyTorchLightning/pytorch-lightning/issues/7166#issuecomment-826808773

**Review by Borda — APPROVED** [review]
